aten/src/ATen/core/intrusive_ptr.h

```diff
@@ -0,0 +1,532 @@
+#pragma once
+
+#include <ATen/Error.h>
+#include <atomic>
+#include <stdexcept>
+
+namespace c10 {
+
+/**
+ * intrusive_ptr<T> is an alternative to shared_ptr<T> that has better
+ * performance because it does the refcounting intrusively
+ * (i.e. in a member of the object itself).
+ * Your class T needs to inherit from intrusive_ptr_target to allow it to be
+ * used in an intrusive_ptr<T>.
+ */
+
+class intrusive_ptr_target {
+  // Note [Weak references for intrusive refcounting]
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // Here's the scheme:
+  //
+  //  - refcount == number of strong references to the object
+  //    weakcount == number of weak references to the object,
+  //      plus one more if refcount > 0
+  //    An invariant: refcount > 0  =>  weakcount > 0
+  //
+  //  - THStorage stays live as long as there are any strong
+  //    or weak pointers to it (weakcount > 0, since strong
+  //    references count as a +1 to weakcount)
+  //
+  //  - finalizers are called and data_ptr is deallocated when refcount == 0
+  //
+  //  - Once refcount == 0, it can never again be > 0 (the transition
+  //    from > 0 to == 0 is monotonic)
+  //
+  //  - When you access THStorage via a weak pointer, you must
+  //    atomically increment the use count, if it is greater than 0.
+  //    If it is not, you must report that the storage is dead.
+  //
+  mutable std::atomic<size_t> refcount_;
+  mutable std::atomic<size_t> weakcount_;
+
+  template <typename T, typename NullType>
+  friend class intrusive_ptr;
+  template <typename T, typename NullType>
+  friend class weak_intrusive_ptr;
+
+ protected:
+  // protected destructor. We never want to destruct intrusive_ptr_target*
+  // directly.
+  virtual ~intrusive_ptr_target() {
+    AT_CHECK(
```

Comment:
Make these AT_ASSERTs, because they indicate internal library error. (There is an important difference: stack traces for AT_CHECK are suppressed, but AT_ASSERT is not). If you still want the message use AT_ASSERTM